### PR TITLE
Improve gas estimate accuracy by enhancements in web3

### LINKED ISSUE #14152: Add integration tests that compare gas estimate results with actual values
### Problem

Due to some reported issues with gas estimate inaccuracy, we need to create more tests that help us find the root causes and verify correctness afterwards.

### Solution

Currently there is a "Gas Estimate Accuracy" workflow that compares gas estimate results with actual gas used of transactions that are executed against the consensus node. We need to implement something similar in the web3 with integration tests. Take the `gas_used` of contract results from testnet/mainnet and use it as an expected value in the tests. Construct the transactions and compare the estimate result with this value. The tests need to compare recent transactions as well as older ones that used old EVM versions and old gas calculators.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by IvanKavaldzhiev**
**Description**:

Fixes gas estimation for contract creates with value and calls that interact with system accounts, and adds fixture-based tests that replay production state. The gas-estimate-accuracy workflow now drops historical transactions that cannot be reproduced at the previous block instead of treating them as estimate failures. These failures pollute the logs and report estimates out of range for simulations that cannot be faithfully reproduced due to block dependent changes. Summary of the changes:

- Set initialBalance on contract-create estimate transactions so value transfers are included.
- Give dummy system accounts a non-zero balance so transfers to accounts such as the staking reward collector estimate correctly.
- Cap binary-search estimates with floor so the 20% maximum stays within the accuracy check’s comparison range.
- Add production-state estimate tests and coverage for the non-empty dummy staking-reward account.
- After an out-of-tolerance estimate, replay the call without estimate and skip the transaction if the result does not match consensus or the original status is CONTRACT_REVERT_EXECUTED.

The size of the attached resources using data prom production is 82 KB.

**Related issue(s)**:

Fixes  #14152

**Notes for reviewer**:

Suggestion for #14152 ticket - it might not be closed with this PR, since additional web3 enhancements might appear in later gas-estimate-accuracy workflow runs. We may take an iterative approach, where issues are polished incrementally by observing the latest runs of the workflow.

With these changes a run against mainnet have been done with 30000 transaction results. Here is the outcome:

  Fetched contract result count: 30000
  Executed validation request count: 8949
  Passed estimation count: 8563
  Out of tolerance estimation count: 146
  Non reproducable contract results: 240
  Skipped request count: 11995
  Reverted estimate request count: 8847
  Errors count outside CONTRACT_REVERT_EXECUTED: 209
  
  However, the contract results used were from transactions with 0.76.1 HAPI version and the hedera-app version used in the web3 module was 0.75.1, which might have affected at least partially the failed results. A previous run with 3000 results reported 0 out of tolerance estimation counts.

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

**Comment by bilyana-gospodinova** [comment]
There are some failing `Non-Zero Realm` tests in CI.

File: web3/src/main/java/org/hiero/mirror/web3/state/keyvalue/AccountReadableKVState.java [hunk 1]
```diff
@@ -85,7 +86,11 @@ public AccountReadableKVState(
     @Override
     protected Account readFromDataSource(@NonNull AccountID key) {
         if (!ContractCallContext.isBalanceCallSafe() && systemAccounts.contains(key)) {
-            return getDummySystemAccountIfApplicable(key).orElse(null);
+            return getDummySystemAccountIfApplicable(key)
+                    .map(account -> account.copyBuilder()
+                            .tinybarBalance(NON_EMPTY_SYSTEM_ACCOUNT_BALANCE)
```
Comment (bilyana-gospodinova) [inline web3/src/main/java/org/hiero/mirror/web3/state/keyvalue/AccountReadableKVState.java:93]: Can you please add a comment in the code why we need this?

File: web3/src/main/java/org/hiero/mirror/web3/service/utils/BinaryGasEstimator.java [hunk 2]
```diff
@@ -60,7 +60,7 @@ public long search(
 
         metricUpdater.accept(totalGasUsed, iterationsMade);
 
-        final long maxAllowedEstimate = (long) Math.ceil(initialGasUsed * 1.20);
+        final long maxAllowedEstimate = (long) Math.floor(initialGasUsed * 1.20);
         if (hi > maxAllowedEstimate) {
             return maxAllowedEstimate;
         }
```
Comment (bilyana-gospodinova) [inline web3/src/main/java/org/hiero/mirror/web3/service/utils/BinaryGasEstimator.java:66]: Can you please add a test in `BinaryGasEstimator` that ensures the returned estimate is always up to 20%?

Such as:
```
@Test
void searchCapDoesNotExceedScriptUpperBound() {
    // 1.20 * gasUsed is not an integer, so ceil() is strictly above the script bound.
    final long gasUsed = 34187L;
    final long high = 15_000_000L;
    final double scriptUpperBound = gasUsed * 1.20;
    final long ceilCap = (long) Math.ceil(scriptUpperBound);
    final long floorCap = (long) Math.floor(scriptUpperBound);
    assertThat(ceilCap).isGreaterThan((long) scriptUpperBound);

    // Fail every probe so the search must return the 20% cap.
    final long estimated = binaryGasEstimator.search(
            (_, _) -> {}, _ -> createTxnResult(gasUsed, false), gasUsed, high);

    assertThat(estimated).isEqualTo(floorCap);
    assertThat((double) estimated).isLessThanOrEqualTo(scriptUpperBound);
}
```

File: web3/src/main/java/org/hiero/mirror/web3/service/utils/BinaryGasEstimator.java [hunk 3]
```diff
@@ -60,7 +60,7 @@ public long search(
 
         metricUpdater.accept(totalGasUsed, iterationsMade);
 
-        final long maxAllowedEstimate = (long) Math.ceil(initialGasUsed * 1.20);
+        final long maxAllowedEstimate = (long) Math.floor(initialGasUsed * 1.20);
```
Comment (bilyana-gospodinova) [inline web3/src/main/java/org/hiero/mirror/web3/service/utils/BinaryGasEstimator.java:63]: Can you please add a test in `BinaryGasEstimator` that ensures the returned estimate is always up to 20%?

Such as:
```
@Test
void searchCapDoesNotExceedScriptUpperBound() {
    // 1.20 * gasUsed is not an integer, so ceil() is strictly above the script bound.
    final long gasUsed = 34187L;
    final long high = 15_000_000L;
    final double scriptUpperBound = gasUsed * 1.20;
    final long ceilCap = (long) Math.ceil(scriptUpperBound);
    final long floorCap = (long) Math.floor(scriptUpperBound);
    assertThat(ceilCap).isGreaterThan((long) scriptUpperBound);

    // Fail every probe so the search must return the 20% cap.
    final long estimated = binaryGasEstimator.search(
            (_, _) -> {}, _ -> createTxnResult(gasUsed, false), gasUsed, high);

    assertThat(estimated).isEqualTo(floorCap);
    assertThat((double) estimated).isLessThanOrEqualTo(scriptUpperBound);
}
```

File: .github/workflows/resources/gas-estimate-accuracy.sh [hunk 4]
```diff
@@ -242,6 +273,14 @@ check_result() {
     break
   done
 
+  local replay_request
+  replay_request="$(jq -c '.estimate = false' <<<"${request}")"
+  post_contracts_call "${replay_request}"
+  if ! is_reproducible_replay "${result_json}" "${http_code}" "${body}"; then
+    nonReproducable=$((nonReproducable + 1))
+    return 0
+  fi
```
Comment (bilyana-gospodinova) [inline .github/workflows/resources/gas-estimate-accuracy.sh:291]: `checked` should be incremented only after this if-statement

File: .github/workflows/resources/gas-estimate-accuracy.sh [hunk 5]
```diff
@@ -35,6 +36,44 @@ hex_to_dec() {
   printf '%d' "0x${hex}"
 }
 
+hex_equal() {
+  local a="${1#0x}" b="${2#0x}"
+  a="${a#0X}"
+  b="${b#0X}"
+  [[ "${a,,}" == "${b,,}" ]]
+}
```
Comment (bilyana-gospodinova) [inline .github/workflows/resources/gas-estimate-accuracy.sh:56]: 0x1 and 0x01 are treated as different with this function. This might cause some false-positives.

File: web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java [hunk 6]
```diff
@@ -1488,4 +1507,295 @@ void processCallWithGasAppliesAccountNonceOverride(final boolean unhex) {
...
+                    ContractCallServiceTest.class.getResourceAsStream("/gas-estimate-accuracy/" + filename)) {
+                return new ObjectMapper().readTree(in);
+            } catch (final Exception e) {
+                throw new ExceptionInInitializerError(e);
+            }
+        }
+
+        private static byte[] decodeHex(final String hex) {
+            return Hex.decode(hex.startsWith(HEX_PREFIX) ? hex.substring(HEX_PREFIX.length()) : hex);
+        }
```
Comment (bilyana-gospodinova) [inline web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java]: Can use `BytesDecoder.hexToBytes()`

File: web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java [hunk 7]
```diff
@@ -1488,4 +1507,295 @@ void processCallWithGasAppliesAccountNonceOverride(final boolean unhex) {
...
+                    .setHbarEquiv(rate.get("hbar_equivalent").asInt())
+                    .setExpirationTime(TimestampSeconds.newBuilder()
+                            .setSeconds(rate.get("expiration_time").asLong()))
+                    .build();
+        }
+
+        private static JsonNode loadFixture(final String filename) {
+            try (final var in =
+                    ContractCallServiceTest.class.getResourceAsStream("/gas-estimate-accuracy/" + filename)) {
+                return new ObjectMapper().readTree(in);
```
Comment (bilyana-gospodinova) [inline web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java]: `new ObjectMapper()` can be static as only 1 instance is needed.

File: web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java [hunk 8]
```diff
@@ -1488,4 +1507,295 @@ void processCallWithGasAppliesAccountNonceOverride(final boolean unhex) {
...
+                    .persist();
+            domainBuilder
+                    .contract()
+                    .customize(c -> c.id(contractId.getId())
+                            .runtimeBytecode(loadHexResource(
+                                    contract.get("runtime_bytecode_file").asText())))
+                    .persist();
+            persistAccountBalance(contractId, balance, stateTimestamp);
+            final var slots = new HashMap<String, String>();
+            copySlots(contract.get("slots"), slots);
```
Comment (bilyana-gospodinova) [inline web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java]: It only copies `JsonNode` fields into a `HashMap` that `persistHistoricalContractSlots` immediately walks again. Can use `contract.path("slots")` to iterate them.

File: web3/src/main/java/org/hiero/mirror/web3/state/keyvalue/AccountReadableKVState.java (hunk 1 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/state/keyvalue/AccountReadableKVState.java:93]: Added

File: web3/src/main/java/org/hiero/mirror/web3/service/utils/BinaryGasEstimator.java (hunk 2 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/service/utils/BinaryGasEstimator.java:66]: Added

File: web3/src/main/java/org/hiero/mirror/web3/service/utils/BinaryGasEstimator.java (hunk 3 above)
Comment (IvanKavaldzhiev) [reply web3/src/main/java/org/hiero/mirror/web3/service/utils/BinaryGasEstimator.java:63]: Added

File: .github/workflows/resources/gas-estimate-accuracy.sh (hunk 4 above)
Comment (IvanKavaldzhiev) [reply .github/workflows/resources/gas-estimate-accuracy.sh:291]: Fixed

File: .github/workflows/resources/gas-estimate-accuracy.sh (hunk 5 above)
Comment (IvanKavaldzhiev) [reply .github/workflows/resources/gas-estimate-accuracy.sh:56]: Fixed

File: web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java (hunk 6 above)
Comment (IvanKavaldzhiev) [reply web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java]: Fixed

File: web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java (hunk 7 above)
Comment (IvanKavaldzhiev) [reply web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java]: Fixed

File: web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java (hunk 8 above)
Comment (IvanKavaldzhiev) [reply web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java]: Fixed

File: .github/workflows/resources/gas-estimate-accuracy.sh [hunk 9]
```diff
@@ -20,6 +20,7 @@ skipped=0
 estimate_reverts=0
 api_errors=0
 page_fetch_failures=0
+nonReproducable=0
```
Comment (bilyana-gospodinova) [inline .github/workflows/resources/gas-estimate-accuracy.sh:23]: nit: nonReproducible

File: web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java [hunk 10]
```diff
@@ -1488,4 +1505,288 @@ void processCallWithGasAppliesAccountNonceOverride(final boolean unhex) {
...
+                            .genesisFeeSchedules(evmProperties.getVersionedConfiguration())
+                            .toByteArray(),
+                    2L);
+            persistSystemFile(systemEntity.simpleFeeScheduleFile(), simpleFeeSchedule, 3L);
+            persistSystemFile(systemEntity.throttleDefinitionFile(), throttleDefinitionBytes(), 4L);
+
+            jdbcTemplate.update("""
+                    update entity
+                    set created_timestamp = ?, timestamp_range = int8range(cast(? as bigint), null)
+                    """, stateTimestamp, stateTimestamp);
```
Comment (bilyana-gospodinova) [inline web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java:1602]: Is this intentional - having an update statement with no "where" clause?

File: web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java [hunk 11]
```diff
@@ -1488,4 +1505,288 @@ void processCallWithGasAppliesAccountNonceOverride(final boolean unhex) {
             assertThat(result.gasUsed()).isPositive();
         }
     }
+
+    @Nested
+    class GasAccuracyEstimateWithProductionState {
```
Comment (bilyana-gospodinova) [inline web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java:1510]: This class can be moved as a separate test suite, instead of nested.

File: web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java (hunk 10 above)
Comment (IvanKavaldzhiev) [reply web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java]: It's intentional, so that all existing records in the entity tables, which are all the system files needed - to receive the needed timestamp at once.

File: .github/workflows/resources/gas-estimate-accuracy.sh (hunk 9 above)
Comment (IvanKavaldzhiev) [reply .github/workflows/resources/gas-estimate-accuracy.sh]: Fixed

File: web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java (hunk 11 above)
Comment (IvanKavaldzhiev) [reply web3/src/test/java/org/hiero/mirror/web3/service/ContractCallServiceTest.java]: I made it as embedded, since it would be hard to scale similar scenarios due to the increase in size of the resource folder, but extracted it for the sake of cleaner code structure.

**Review by bilyana-gospodinova — APPROVED** [review]

**Review by steven-sheehy — APPROVED** [review]
